# <div class="alert alert-block alert-info"> <span style="font-family: Times New Roman">Linear Algebra </span> </div>

Linear Algebra is a branch of mathematics that is extremely useful in data science and machine learning. Linear algebra is the most important math skill in machine learning. Most machine learning models can be expressed in matrix form. A dataset itself is often represented as a matrix. Linear algebra is used in data preprocessing, data transformation, and model evaluation.

- **Solving equations**: Data scientists often use algebra to solve complex equations, which can be used to predict future trends, identify patterns, and make informed decisions.

- **Linear algebra**: Linear algebra is used to analyze linear equations and systems of equations, which are commonly found in data science. Linear algebra is also used to perform operations such as matrix multiplication and inversion, which are fundamental to many machine learning algorithms.

- **Probability and statistics**: Algebra is used to understand the principles of probability and statistics, which are essential to data science. These principles are used to interpret data, draw conclusions, and make predictions.

- **Data transformation**: Algebraic operations such as scaling, translation, and rotation are used to transform data into a more meaningful format. These transformations help in data visualization, clustering, and classification.

- **Optimization**: Algebra is used to optimize mathematical models and algorithms. This helps data scientists to find the best possible solution to a problem, given a set of constraints.

__You can use numpy, scipy for the following examples__

__Q1. Given Matrix A below, calculate the eigen values of the matrix using Python.__

In [ ]:
import numpy as np

# Define the matrix A
A = np.array([[2, 1, 0], [1, 2, 1], [0, 1, 2]])

# Calculate the eigenvalues of A
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvalues:", eigenvalues)


Eigenvalues: [3.41421356 2.         0.58578644]


So the eigenvalues of matrix A are 3.41421356, 2, and 0.58578644.

__Q2. Given Matrix A below, calculate the eigen vectors of the matrix using Python.__

In [ ]:
# Define the matrix A
A = np.array([[5, 6], [9, 3]])

# Calculate the eigenvectors of A
eigenvalues, eigenvectors = np.linalg.eig(A)

print("Eigenvectors:")
for i in range(len(eigenvectors)):
    print(eigenvectors[:,i])


Eigenvectors:
[0.68302051 0.73039919]
[-0.58049671  0.81426259]


__Q3. Given Matrix A above, calculate the transpose of the matrix using numpy and without using numpy.__

In [ ]:
# Define the matrix A
A = np.array([[7, 7, 3], [2, 9, 6], [0, 5, 1]])

print("Matrix\n", A)
# Calculate the transpose of A using NumPy
A_transpose = np.transpose(A)

print("\nTranspose of A using NumPy:\n", A_transpose)


Matrix
 [[7 7 3]
 [2 9 6]
 [0 5 1]]

Transpose of A using NumPy:
 [[7 2 0]
 [7 9 5]
 [3 6 1]]


**Without using NumPy**

In [ ]:
# define the matrix A
A = [[7, 7, 3], [2, 9, 6], [0, 5, 1]]

# define an empty matrix B with switched dimensions
B = [[0, 0, 0], [0, 0, 0], [0, 0, 0]]

# iterate through the rows of A
for i in range(len(A)):
    # iterate through the columns of A
    for j in range(len(A[0])):
        # switch the rows and columns and store the result in B
        B[j][i] = A[i][j]

# print the transpose matrix B
print(B)

[[7, 2, 0], [7, 9, 5], [3, 6, 1]]


__Q4. Given Matrix A above, calculate the trace of the matrix using numpy and without using numpy.__

**Using NumPy**

In [ ]:
# Define the matrix A
A = np.array([[7, 7, 3], [2, 9, 6], [0, 5, 1]])

# Calculate the trace of A using NumPy
A_trace = np.trace(A)

print("Trace of A using NumPy:", A_trace)


Trace of A using NumPy: 17


**Without using NumPy**

In [ ]:
# Define the matrix A
A = [[7, 7, 3], [2, 9, 6], [0, 5, 1]]

# Calculate the trace of A without using NumPy
A_trace = 0
for i in range(len(A)):
    A_trace += A[i][i]

print("Trace of A without using NumPy:", A_trace)


Trace of A without using NumPy: 17


__Q5. Given Matrix A above, calculate the inverse of the matrix using numpy.__

In [ ]:
# Define the matrix A
A = np.array([[7, 7, 3], [2, 9, 6], [0, 5, 1]])

# Calculate the inverse of A using NumPy
A_inv = np.linalg.inv(A)

print("Inverse of A using NumPy:\n", A_inv)


Inverse of A using NumPy:
 [[ 0.16030534 -0.0610687  -0.11450382]
 [ 0.01526718 -0.05343511  0.27480916]
 [-0.07633588  0.26717557 -0.3740458 ]]


__Q6. Given Matrix A above, calculate the determinant of the matrix using numpy.__

In [ ]:
# Define the matrix A
A = np.array([[7, 7, 3], [2, 9, 6], [0, 5, 1]])

# Calculate the determinant of A using NumPy
A_det = np.linalg.det(A)

print("Determinant of A using NumPy:", A_det)


Determinant of A using NumPy: -130.99999999999994


__Q7. Given Matrix A,B below, calculate the dot product of the matrices using numpy and without using numpy.__

In [ ]:
# Define the matrices A and B
A = np.array([[35, 46, 67], [33, 42, 65], [24, 23, 51]])
B = np.array([[23, 43, 54], [43, 44, 87], [12, 43, 44]])

# Calculate the dot product of A and B using NumPy
C = np.dot(A, B)

print("Dot product of A and B using NumPy:\n", C)


Dot product of A and B using NumPy:
 [[3587 6410 8840]
 [3345 6062 8296]
 [2153 4237 5541]]


Whithout using library

In [ ]:
# Define a function to calculate dot product of two matrices
def dot_product(A, B):
    C = [[0 for _ in range(len(B[0]))] for _ in range(len(A))]
    for i in range(len(A)):
        for j in range(len(B[0])):
            for k in range(len(B)):
                C[i][j] += A[i][k] * B[k][j]
    return C

# Define the matrices A and B
A = [[35, 46, 67], [33, 42, 65], [24, 23, 51]]
B = [[23, 43, 54], [43, 44, 87], [12, 43, 44]]

# Calculate the dot product of A and B using the defined function
C = dot_product(A, B)

print("Dot product of A and B without NumPy:\n", C)


Dot product of A and B without NumPy:
 [[3587, 6410, 8840], [3345, 6062, 8296], [2153, 4237, 5541]]


__Q8. Given Matrix A,B above, calculate the cross product of the matrices using numpy and without using numpy.__

In [ ]:
A = np.array([[35, 46, 67],
              [33, 42, 65],
              [24, 23, 51]])

B = np.array([[23, 43, 54],
              [43, 44, 87],
              [12, 43, 44]])

C = np.multiply(A, B)  # element-wise multiplication
print(C)


[[ 805 1978 3618]
 [1419 1848 5655]
 [ 288  989 2244]]


Without using library

In [ ]:
A = [[35, 46, 67],
     [33, 42, 65],
     [24, 23, 51]]

B = [[23, 43, 54],
     [43, 44, 87],
     [12, 43, 44]]

C = [[0, 0, 0],
     [0, 0, 0],
     [0, 0, 0]]

for i in range(len(A)):
    for j in range(len(B)):
        C[i][j] = A[i][j] * B[i][j]

print(C)


[[805, 1978, 3618], [1419, 1848, 5655], [288, 989, 2244]]


__Q9. Given two vectors a,b calculate the inner product of the vectors using numpy.__

In [ ]:
a = np.array([1, 2, 5])
b = np.array([2, 1, 0])

inner_product = np.dot(a, b)
print("Inner product of the said vectors:\n",inner_product)


Inner product of the said vectors:
 4


__Q10. Given two matrices x,y calculate the inner product of the matrices using numpy.__

In [ ]:
x = np.array([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
y = np.array([[3, 4, 5], [6, 7, 8], [9, 10, 11]])

inner_product = np.matmul(x, y)
print("Inner product of the said vectors:\n",inner_product)


Inner product of the said vectors:
 [[ 24  27  30]
 [ 78  90 102]
 [132 153 174]]


__Q.11 Given two matrices A,B calculate the multiplication using numpy.__

In [ ]:
A = np.array([[1, 0], [0, 1]])
B = np.array([[1, 2], [3, 4]])

C = np.matmul(A, B)
print("Result of the said matrix multiplication:\n",C)

Result of the said matrix multiplication:
 [[1 2]
 [3 4]]


__Q.12 Given two matrices A,B calculate the outer product using numpy.__

In [ ]:
import numpy as np

A = np.array([[1, 0], [0, 1]])
B = np.array([[1, 2], [3, 4]])

C = np.outer(A, B)
print("Outer product of the said two vectors:\n",C)


Outer product of the said two vectors:
 [[1 2 3 4]
 [0 0 0 0]
 [0 0 0 0]
 [1 2 3 4]]


__Q.13 Given a vector, calculate the normalized form of a vector using numpy.__

In [ ]:
A = np.array([[1, 2], [3, 4]])

# Calculate the magnitude of the vector
magnitude = np.linalg.norm(A)

# Normalize the vector by dividing each element by the magnitude
normalized_vector = A / magnitude

print(normalized_vector)

[[0.18257419 0.36514837]
 [0.54772256 0.73029674]]


In [ ]:
vector_norm = np.linalg.norm(A)

print("Vector norm: ", vector_norm)


Vector norm:  5.477225575051661


__Q.14 Given a matrix, calculate the normalized form of a matrix using numpy.__

In [ ]:
import numpy as np

A = np.array([[1, 2], [3, 4], [5, 6]])

matrix_norm = np.linalg.norm(A)

print("Matrix norm: ", matrix_norm)


Matrix norm:  9.539392014169456


In [ ]:
# define the matrix
A = np.array([[1, 2, 3], [4, 5, 6], [7, 8, 9]])

# calculate the norm of the matrix
norm_A = np.linalg.norm(A)

# normalize the matrix
A_norm = A / norm_A

print("Original matrix:\n", A)
print("Normalized matrix:\n", A_norm)


Original matrix:
 [[1 2 3]
 [4 5 6]
 [7 8 9]]
Normalized matrix:
 [[0.05923489 0.11846978 0.17770466]
 [0.23693955 0.29617444 0.35540933]
 [0.41464421 0.4738791  0.53311399]]



# <div class="alert alert-block alert-info"> <span  style= "font-family: Times New Roman">Probability    </span> </div>

Probability is the branch of mathematics that deals with the study of random events and the likelihood of their occurrence. In data science, probability is a fundamental concept that is used extensively in many areas. Here are some reasons why probability is important for data science:

- **Modeling uncertainty**: Data science deals with uncertain events and probability helps to model and quantify that uncertainty. Probability helps to understand the randomness in the data and provide a framework for making predictions.

- **Statistical inference**: Probability is the foundation of statistical inference, which is the process of drawing conclusions about a population based on a sample of data. Probability distributions are used to model the variation in the data and make inferences about the population.

- **Machine learning**: Machine learning algorithms use probability to make predictions and classify data. Probability models, such as Bayesian networks and Hidden Markov Models, are used to represent uncertainty in the data and make predictions.

- **Decision-making**: Probability is used in decision-making to evaluate the risks and benefits of different options. In data science, probability is used to make decisions based on data and to evaluate the effectiveness of different strategies.

- **Data cleaning**: Probability is used in data cleaning to identify outliers and anomalies in the data. Probability models, such as Gaussian distributions, are used to identify data points that are unlikely to occur by chance.

<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
1. What is the probability of drawing two cards (from the same deck of cards) that have the same suite?<br>
</p>
</span>
</div>

### Solution: 
This is an example of a dependent event. The probability that two events will occur in the case of dependent event can be defined as:

$ P(A\cap B)=P(A)*P(B|A) $ <br>
$ P(A∩B)=P(A)∗P(B∣A) $ <br>
which means that the probability of event A and event B to both occur is the same as the probability of event A to occur times the probability of event B to occur given the result of event A. <br>

In our case, there are four suites in a deck of cards, and each suite has 13 cards. <br>

In the first draw, our probability to get a card with a specific suite would be 13/52. In the second draw, our probability of getting a card with the same suite as the first one wouldn’t be 13/52 anymore, but 12/51. <br>
Hence: <br>

$ P(two\,cards\,same\,suite) = 4*\frac{13}{52}*\frac{12}{51}=\frac{4}{17} $ <br>


<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
2. What is the probability of choosing 2 queens out of a deck of cards?<br>
</p>
</span>
</div>

### Solutions: 
This is also an example of a dependent event. In the first draw, our probability of getting a queen is 4/52. If we do get a queen in the first draw, then our probability to get another queen in the second draw would be 3/51. Hence:

$ P(2\,queens) = P(X_1=queen)\cap P(X_2=queen) $<br>
$P(2\,queens) = \frac{1}{13}*\frac{1}{17}$ <br>
$P(2\,queens) = \frac{1}{221} $<br>


<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
3. Let’s say you have 2 dice. What is the probability of getting at least one 4?<br>
</p>
</span>
</div>

### Solutions:
Let’s say that: <br>

A = getting a 4 in the first die <br>
B = getting a 4 in the second die <br>

The probability of independent events A and B both to occur can be defined as: <br>

$ P(A\cap B)=P(A)*P(B) $ <br>
$ P(A∩B)=P(A)∗P(B)$ <br>
And the probability of getting at least one 4 can be computed with the probability of union of two events: <br>

$ P(A\cup B)=P(A)+P(B)-P(A\cap B)$ <br>
$ P(A∪B)=P(A)+P(B)−P(A∩B)$ <br>
We know that the probability of us getting any specific outcome from throwing a die is ⅙. Thus, <br>

$P(A\cup B)=\frac{1}{6}+\frac{1}{6}-(\frac{1}{6}*\frac{1}{6})=\frac{11}{36}$ <br>


<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
4. Three ants are sitting at the three corners of an equilateral triangle. Each ant randomly picks a direction and starts to move along the edge of the triangle. What is the probability that none of the ants collide?”<br>
</p>
</span>
</div>

### Solution:
this is the case of an independent event. Each ant can randomly pick the direction, either to the left or to the right. The decision of one ant to go to the left wouldn’t affect the decision of the other two ants whether they want to go to the left or right. <br>

Since the decision is random, then the probability of an ant to pick a certain direction is 0.5. The three ants wouldn’t collide if all of them go to the left or all of them go to the right. <br>

Hence: <br>

$ P(none\,of\,the\,ants\,collide)=(\frac{1}{2})^3+(\frac{1}{2})^3=\frac{1}{4} $<br>


<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
5. Consider a tetrahedral die and roll it twice. What is the probability that the number on the first roll is strictly higher than the number on the second roll? <br>

Note: A tetrahedral die has only four sides (1, 2, 3 and 4). <br>
</p>
</span>
</div>

### Solution: 
(1,1) (2,1) (3,1) (4,1) 
(1,2) (2,2) (3,2) (4,2) 
(1,3) (2,3) (3,3) (4,3)
(1,4) (2,4) (3,4) (4,4) <br>
There are 6 out of 16 possibilities where the first roll is strictly higher than the second roll.

Answer: 3/8


<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
6. A fair six-sided die is rolled 6 times. What is the probability of getting all outcomes as unique? <br>
</p>
</span>
</div>

### Solution:
For all the outcomes to be unique, we have 6 choices for the first turn, 5 for the second turn, 4 for the third turn and so on

Therefore the probability if getting all unique outcomes will be equal to 0.01543

Answer: 0.01543

<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
7. We have two coins, A and B. For each toss of coin A, the probability of getting head is 1/2 and for each toss of coin B, the probability of getting Heads is 1/3. All tosses of the same coin are independent. We select a coin at random and toss it till we get a head. The probability of selecting coin A is ¼ and coin B is 3/4. What is the expected number of tosses to get the first heads?<br>
</p>
</span>
</div>

### Solution:
If coin A is selected then the number of times the coin would be tossed for a guaranteed Heads is 2, similarly, for coin B it is 3. Thus the number of times would be

Tosses = 2 * (1/4)[probability of selecting coin A] + 3*(3/4)[probability of selecting coin B]

             = 2.75
             
          Answer: 2.75

 

<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
8. Suppose a life insurance company sells a \$240000 one year term life insurance policy to a 25-year old female for $210. The probability that the female survives the year is .999592. Find the expected value of this policy for the insurance company.<br>
</p>
</span>
</div>

### Solution:
P(company loses the money ) = 0.99592

P(company does not lose the money ) = 0.000408

The amount of money company loses if it loses = 240,000 – 210 = 239790

While the money it gains is $210

Expected money the company will have to give = 239790*0.000408 = 97.8

Answer: 97.8

<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
9.Some test scores follow a normal distribution with a mean of 18 and a standard deviation of 6. What proportion of test takers have scored between 18 and 24?<br>
</p>
</span>
</div>

### Solution:

So here we would need to calculate the Z scores for value being 18 and 24. We can easily doing that by putting sample mean as 18 and population mean as 18 with σ = 6 and calculating Z. Similarly we can calculate Z for sample mean as 24.

Z= (X-μ)/σ

Therefore for 26 as X,

Z = (18-18)/6  = 0 , looking at the Z table we find 50% people have scores below 18.

For 24 as X

Z = (24-18)/6  = 1, looking at the Z table we find 84% people have scores below 24.

Therefore around 34% people have scores between 18 and 24.

<div class='warning' style='padding:0.1em; background-color:#E9D8FD; color:#69337A'>
<span>
<p style='margin-left:1em;'>
10.Suppose you were interviewed for a technical role. 50% of the people who sat for the first interview received the call for second interview. 95% of the people who got a call for second interview felt good about their first interview. 75% of people who did not receive a second call, also felt good about their first interview. If you felt good after your first interview, what is the probability that you will receive a second interview call?<br>
</p>
</span>
</div>

### Solution:
Answer: 75%

Solve it using Bayes theorem 

**Health Insurance**

Health insurance companies often use probability to determine how likely it is that certain individuals will spend a certain amount on healthcare each year.

For example, a company might use factors like age, existing medical conditions, current health status, etc. to determine that there’s a 80% probability that a certain individual will spend $8,000 or more on healthcare in a given year.

Individuals who are likely to spend more on healthcare will be charged higher premiums because the insurance company knows that they’ll be more expensive to insure.

In [ ]:
import scipy.stats as stats

# Define the variables
mu = 8000  # mean
sigma = 2000  # standard deviation

# Calculate the z-score for the given probability
p = 0.80
z = stats.norm.ppf(p)

# Calculate the healthcare cost using the inverse transform method
x = mu + z * sigma

print(f"There is an 80% probability that the healthcare cost will be {x:.2f} or more.")


There is an 80% probability that the healthcare cost will be 9683.24 or more.


In [ ]:
# There is an 80% probability that the healthcare cost will be 10045.38 or more.


This means that there is an 80% probability that the healthcare cost for this individual will be $10,045.38 or more, given their age, existing medical conditions, and current health status.


### **Happy Learning 🙂**